### ISSUE TITLE
Zoltan2: Conditional jump depends on unitialized value

### ISSUE THREAD

**Issue by cgcgcg** [trilinos/Trilinos#6399]
## Bug Report
@trilinos/zoltan2 

### Description
@csiefer2 ran EMPIRE with Valgrind and found this:
```
==15953== Conditional jump or move depends on uninitialised value(s)
==15953== at 0xE91927E:
Kokkos::Impl::ParallelReduce<Zoltan2::ReduceWeightsFunctor<Kokkos::TeamPolicy<Kokkos::OpenMP>,
double, int, int, Kokkos::Device<Kokkos::OpenMP, Kokkos::HostSpace>,
double>, Kokkos::TeamPolicy<Kokkos::OpenMP>, Kokkos::InvalidType,
Kokkos::OpenMP>::execute() const [clone ._omp_fn.45] (in
/home/caglusa/Trilinos-install/lib/libmuelu.so.12.17)
==15953==    by 0x26FD8514: GOMP_parallel (parallel.c:168)
==15953== by 0xE9571B2: Zoltan2::AlgMJ<double, int, long long, int,
Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP,
Kokkos::HostSpace> >::mj_1D_part_get_part_weights(int, int,
Kokkos::View<double*, Kokkos::Device<Kokkos::OpenMP,
Kokkos::HostSpace> >&, int) (in
/home/caglusa/Trilinos-install/lib/libmuelu.so.12.17)
==15953== by 0xE96645B: Zoltan2::AlgMJ<double, int, long long, int,
Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP,
Kokkos::HostSpace> >::mj_1D_part(Kokkos::View<double*,
Kokkos::Device<Kokkos::OpenMP, Kokkos::HostSpace> >&, double, int,
int, Kokkos::View<double*, Kokkos::Device<Kokkos::OpenMP,
Kokkos::HostSpace> >&, int, Kokkos::View<int*,
Kokkos::Device<Kokkos::OpenMP, Kokkos::HostSpace> >&,
Kokkos::View<unsigned long*, Kokkos::Device<Kokkos::OpenMP,
Kokkos::HostSpace> >&) (in
/home/caglusa/Trilinos-install/lib/libmuelu.so.12.17)
==15953== by 0xE97FFEB: Zoltan2::AlgMJ<double, int, long long, int,
Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP,
Kokkos::HostSpace>
>::multi_jagged_part(Teuchos::RCP<Zoltan2::Environment const> const&,
Teuchos::RCP<Teuchos::Comm<int> const>&, double, int, unsigned long,
Kokkos::View<int*, Kokkos::HostSpace>&, int, int, int, long long,
Kokkos::View<long long const*, Kokkos::Device<Kokkos::OpenMP,
Kokkos::HostSpace> >&, Kokkos::View<double**, Kokkos::LayoutLeft,
Kokkos::Device<Kokkos::OpenMP, Kokkos::HostSpace> >&, int,
Kokkos::View<bool*, Kokkos::HostSpace>&, Kokkos::View<double**,
Kokkos::Device<Kokkos::OpenMP, Kokkos::HostSpace> >&,
Kokkos::View<bool*, Kokkos::HostSpace>&, Kokkos::View<int*,
Kokkos::Device<Kokkos::OpenMP, Kokkos::HostSpace> >&,
Kokkos::View<long long*, Kokkos::Device<Kokkos::OpenMP,
Kokkos::HostSpace> >&) (in
/home/caglusa/Trilinos-install/lib/libmuelu.so.12.17)
==15953== by 0xE9870F3:
Zoltan2::Zoltan2_AlgMJ<Zoltan2::XpetraMultiVectorAdapter<Xpetra::MultiVector<double,
int, long long,
Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP,
Kokkos::HostSpace> > >
>::partition(Teuchos::RCP<Zoltan2::PartitioningSolution<Zoltan2::XpetraMultiVectorAdapter<Xpetra::MultiVector<double,
int, long long,
Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP,
Kokkos::HostSpace> > > > > const&) (in
/home/caglusa/Trilinos-install/lib/libmuelu.so.12.17)
==15953== by 0xE93F86D:
Zoltan2::PartitioningProblem<Zoltan2::XpetraMultiVectorAdapter<Xpetra::MultiVector<double,
int, long long,
Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP,
Kokkos::HostSpace> > > >::solve(bool) (in
/home/caglusa/Trilinos-install/lib/libmuelu.so.12.17)
==15953== by 0xE99719E: MueLu::Zoltan2Interface<double, int, long
long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP,
Kokkos::HostSpace> >::Build(MueLu::Level&) const (in
/home/caglusa/Trilinos-install/lib/libmuelu.so.12.17)
==15953==    by 0xDA038EB: MueLu::SingleLevelFactoryBase::CallBuild(MueLu::Level&) const (in /home/caglusa/Trilinos-install/lib/libmuelu-adapters.so.12.17)
==15953== by 0xE5D968F: Teuchos::RCP<Xpetra::Vector<long long, int,
long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP,
Kokkos::HostSpace> > >&
MueLu::Level::Get<Teuchos::RCP<Xpetra::Vector<long long, int, long
long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP,
Kokkos::HostSpace> > > >(std::__cxx11::basic_string<char,
std::char_traits<char>, std::allocator<char> > const&,
MueLu::FactoryBase const*) (in
/home/caglusa/Trilinos-install/lib/libmuelu.so.12.17)
==15953== by 0xE7FA4FB: MueLu::RepartitionFactory<double, int, long
long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP,
Kokkos::HostSpace> >::Build(MueLu::Level&) const (in
/home/caglusa/Trilinos-install/lib/libmuelu.so.12.17)
==15953==    by 0xD9F24B5: MueLu::RefMaxwell<double, int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP, Kokkos::HostSpace> >::compute(bool) (in /home/caglusa/Trilinos-install/lib/libmuelu-adapters.so.12.17)

```

Not sure if this leads to any issues or not.

**Comment by MicheldeMessieres** [comment]
I'll check into this now.

**Comment by MicheldeMessieres** [comment]
@cgcgcg PR #6406 will resolve the valgrind warning. This fix should not have any impact on the output.
@csiefer2 Thanks for spotting this.

**Comment by cgcgcg** [comment]
@MicheldeMessieres Should we keep this issue open as a reminder that we did not understand how a benign change could fix a bug? Otherwise feel free to close.

**Comment by MicheldeMessieres** [comment]
@cgcgcg @mhoemmen Mark had indicated all places where MPI takes a buffer should be HostSpace and not CudaUVMSpace so this change would not be benign. This bug fix also fixed #6412, likely because one of the buffers was explicitly using CudaSpace for UVM testing so there was a 2nd error resolved.  Issue #6422 will be resolved by similar changes which I am working on now. I tried just changing all mirrors to HostSpace and it cleared that bug. Now verifying if I can still clear the bug just changing the MPI calls. So I think we can close this issue.

**Comment by MicheldeMessieres** [comment]
@cgcgcg Sorry deleted my previous. I was thinking I was on a different issue. I think we can close this and I think the issue is understood. I pad arrays with one element on either side to avoid an extra if statement in kernels, to reduce branching. The element was not initialized so we end up doing > < checks on the element and it's never used. I initialized it so valgrind would not complain but there should be no side effects either way.
